File: src/javascript/algorithms/array/11.rotateArray/rotateArray.js
// 1. Rotate Array in right direction Using Reversal Algorithm (Efficient O(n) time, O(1) space)
function rotate(nums, n) {
    const length = nums.length;
    if (length === 0 || n % length === 0) return nums; // No rotation needed

    n = n % length; // Handle n > length

    // Reverse entire array
    reverse(nums, 0, length - 1);
    // Reverse first part
    reverse(nums, 0, n - 1);
    // Reverse second part
    reverse(nums, n, length - 1);

    return nums;
}

// Helper function to reverse elements between two indices
function reverse(arr, start, end) {
    while (start < end) {
        [arr[start], arr[end]] = [arr[end], arr[start]];
        start++;
        end--;
    }
}

// 2. Rotate Array Using Brute Force with pop() and unshift() (Inefficient: O(n^2) time)
function rotateBruteForce(nums, n) {
    const length = nums.length;
    if (length === 0) return;

    n = n % length;
    for (let i = 0; i < n; i++) {
        nums.unshift(nums.pop());
    }
}

// 3. Rotate Array in left direction Using Reversal Algorithm (Efficient O(n) time, O(1) space)
function rotateLeft(nums, n) {
    const length = nums.length;
    if (length === 0 || n % length === 0) return nums; // No rotation needed

    n = n % length; // Handle n > length

    // Reverse first part
    reverse(nums, 0, n - 1);
    // Reverse second part
    reverse(nums, n, length - 1);
    // Reverse entire array
    reverse(nums, 0, length - 1);

    return nums;
}

// 4. Rotate Array left Using Brute Force with shift() and push() (Inefficient: O(n^2) time)
function rotateBruteForceLeft(nums, n) {
    const length = nums.length;
    if (length === 0) return;

    n = n % length;
    for (let i = 0; i < n; i++) {
        nums.push(nums.shift());
    }
}

// ----------------------------
// Test Cases
// ----------------------------
let rotate1 = [1, 2, 3, 4, 5, 6, 7];
console.log("Before rotate (reversal):", [...rotate1]);
rotate(rotate1, 4);
console.log("After rotate (reversal):", rotate1);
rotateLeft(rotate1, 4);

let rotate2 = [-10, 4, 5, -1];
console.log("Before rotate (brute force):", [...rotate2]);
rotateBruteForce(rotate2, 2);
console.log("After rotate (brute force):", rotate2);
rotateBruteForceLeft(rotate2, 2);

// Edge case
let rotate3 = [1];
console.log("Before rotate (single element):", [...rotate3]);
rotate(rotate3, 10);
console.log("After rotate (single element):", rotate3);
rotateLeft(rotate3, 10);

File: src/python/algorithms/array/11.rotateArray/rotate_array.py
"""
Rotate Array

Given an integer array nums, rotate the array to the right by k steps.

Time Complexity: O(n)
Space Complexity: O(1)
"""


def rotate_right(nums, k):
    """
    Rotate array right using reversal algorithm.
    TC: O(n), SC: O(1)
    """
    length = len(nums)
    if length == 0 or k % length == 0:
        return nums
    
    k = k % length  # Handle k > length
    
    # Reverse entire array
    reverse(nums, 0, length - 1)
    # Reverse first part
    reverse(nums, 0, k - 1)
    # Reverse second part
    reverse(nums, k, length - 1)
    
    return nums


def rotate_left(nums, k):
    """
    Rotate array left using reversal algorithm.
    TC: O(n), SC: O(1)
    """
    length = len(nums)
    if length == 0 or k % length == 0:
        return nums
    
    k = k % length  # Handle k > length
    
    # Reverse first part
    reverse(nums, 0, k - 1)
    # Reverse second part
    reverse(nums, k, length - 1)
    # Reverse entire array
    reverse(nums, 0, length - 1)
    
    return nums


def reverse(arr, start, end):
    """Helper function to reverse elements between two indices."""
    while start < end:
        arr[start], arr[end] = arr[end], arr[start]
        start += 1
        end -= 1


def rotate_using_array(nums, k):
    """
    Rotate using additional array.
    TC: O(n), SC: O(n)
    """
    length = len(nums)
    if length == 0:
        return nums
    
    k = k % length
    result = [0] * length
    
    for i in range(length):
        result[(i + k) % length] = nums[i]
    
    return result


# Test cases
if __name__ == "__main__":
    # Test rotate right
    nums1 = [1, 2, 3, 4, 5, 6, 7]
    k1 = 3
    print(f"Original: {nums1}, k={k1}")
    print(f"Rotated right: {rotate_right(nums1.copy(), k1)}")  # [5, 6, 7, 1, 2, 3, 4]
    
    # Test rotate left
    nums2 = [1, 2, 3, 4, 5, 6, 7]
    k2 = 3
    print(f"Original: {nums2}, k={k2}")
    print(f"Rotated left: {rotate_left(nums2.copy(), k2)}")  # [4, 5, 6, 7, 1, 2, 3]
    
    # Test rotate using array
    nums3 = [1, 2, 3, 4, 5, 6, 7]
    k3 = 3
    print(f"Original: {nums3}, k={k3}")
    print(f"Rotated (using array): {rotate_using_array(nums3, k3)}")
